lsp-mode.el

```diff
@@ -903,6 +946,232 @@ WORKSPACE is the workspace that contains the diagnostics."
                                                            (_ :note))
                                                          message)))))))
 
+(defun lsp--ht-get (tbl &rest keys)
+  "Get nested KEYS in TBL."
+  (let ((val tbl))
+    (while (and keys val)
+      (setq val (ht-get val (first keys)))
+      (setq keys (rest keys)))
+    val))
+
+;; lenses support
+
+(defun lsp--lens-text-width (from to)
+  "Measure the width of the text between FROM and TO.
+Results are meaningful only if FROM and TO are on the same line."
+  ;; `current-column' takes prettification into account
+  (- (save-excursion (goto-char to) (current-column))
+     (save-excursion (goto-char from) (current-column))))
+
+(defun lsp--lens-update (ov)
+  "Redraw quick-peek overlay OV."
+  (let ((offset (lsp--lens-text-width (save-excursion
+                                          (beginning-of-visual-line)
+                                          (point))
+                                        (save-excursion
+                                          (beginning-of-line-text)
+                                          (point)))))
+    (save-excursion
+      (goto-char (overlay-start ov))
+      (overlay-put ov
+                   'before-string
+                   (concat (make-string offset ?\s)
+                           (overlay-get ov 'lsp--lens-contents)
+                           "\n")))))
+
+(defun lsp--lens-overlay-ensure-at (pos)
+  "Find or create a lens for the line at POS."
+  (or (car (cl-remove-if-not (lambda (ov) (lsp--lens-overlay-matches-pos ov pos)) lsp--lens-overlays))
+      (let* ((ov (save-excursion
+                   (goto-char pos)
+                   (make-overlay (point-at-bol) (1+ (point-at-eol))))))
```

Comment:
Set front-advance but unset rear-advance

Reply:
Can you elaborate?

Reply:
`(make-overlay (point-at-bol) (1+ (point-at-eol) nil 'front-advance)`